synapse/handlers/abuse_reports.py

```diff
@@ -0,0 +1,372 @@
+    # circular dependency.
+    from synapse.server import HomeServer
+
+from synapse.api.auth import Auth
+from synapse.api.constants import EventTypes
+from synapse.api.errors import Codes, LimitExceededError, SynapseError
+from synapse.api.ratelimiting import Ratelimiter
+from synapse.federation.sender import AbstractFederationSender
+from synapse.http.servlet import assert_params_in_dict
+from synapse.http.site import SynapseRequest
+from synapse.server_notices.server_notices_manager import ServerNoticesManager
+from synapse.state import StateHandler
+from synapse.storage import DataStore, Storage
+from synapse.types import (
+    EventID,
+    JsonDict,
+    RoomID,
+    UserID,
+    create_requester,
+    get_domain_from_id,
+)
+from synapse.visibility import filter_events_for_client
+
+logger = logging.getLogger(__name__)
+
+
+class AbuseReportHandler:
+    auth: Auth
+    federation_sender: Optional[AbstractFederationSender]
+    server_notices: ServerNoticesManager
+    state: StateHandler
+    storage: Storage
+    store: DataStore
+
+    def __init__(self, hs: "HomeServer"):
+        self.hs = hs
+        self.auth = hs.get_auth()
+        self.config = hs.get_config()
+
+        self.federation_sender = None
+        if hs.should_send_federation():
+            self.federation_sender = hs.get_federation_sender()
+        hs.get_federation_registry().register_edu_handler(
+            "org.matrix.m.content_report", self.on_receive_report_through_federation
+        )
+        self.clock = hs.get_clock()
+        self.federation_client = hs.get_federation_client()
+        self.server_notices = hs.get_server_notices_manager()
+        self.state = hs.get_state_handler()
+        self.store = hs.get_datastore()
+        self.storage = hs.get_storage()
+
+        # A rate limiter for incoming room key requests per origin.
+        self._room_key_request_rate_limiter = Ratelimiter(
+            store=hs.get_datastore(),
+            clock=self.clock,
+            rate_hz=self.config.rc_message.per_second,
+            burst_count=self.config.rc_message.burst_count,
+        )
+
+    async def report(
+        self,
+        request: SynapseRequest,
+        user_id: UserID,
+        body: dict,
+        room_id_: str,
+        event_id_: str,
+    ):
+        # Create a requester for the rate limiter.
+        requester = create_requester(user_id)
+        time_now_s = self.clock.time()
+
+        (
+            allowed,
+            time_allowed,
+        ) = await self._room_key_request_rate_limiter.can_do_action(requester=requester)
+        if not allowed:
+            raise LimitExceededError(
+                retry_after_ms=int(1000 * (time_allowed - time_now_s))
+            )
+
+        room_id = RoomID.from_string(room_id_)
+        event_id = None
+        try:
+            event_id = EventID.from_string(event_id_)
+        except Exception:
+            event_id = EventID.from_string("%s:%s" % (event_id_, self.hs.hostname))
```

Comment:
this isn't a thing we do anywhere else - event ids should be treated as complete, opaque, identifiers.

Reply:
Yeah, I was misled by the existence of `EventID`.